### PR TITLE
Support AWS Fargate

### PR DESCRIPTION
**Pull request by nishigori**
Hope to supporting launch_type `"FARGATE"`

* this proposal is non-breaking changes for launch_type: EC2 (=> v0.4.0)

### CODE REVIEW

File: service_load_balancing/variables.tf [hunk 1]
```diff
@@ -57,41 +71,37 @@ variable "iam_role_inline_policy" {
...
 }
 
-variable "deployment_maximum_percent" {
-  description = "The upper limit of the number of running tasks that can be running in a service during a deployment"
-  default     = 200
+# ECS launch_type: EC2
+
+variable "cluster_id" {
+  description = "ID (ARN) of ECS Cluster if created by external"
+  default     = ""
```
Comment (tamakiii) [inline service_load_balancing/variables.tf]: `aws_ecs_service.cluster` is optional (just note)
https://www.terraform.io/docs/providers/aws/r/ecs_service.html#cluster

File: service_load_balancing/launch_type_ec2.tf [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+  # As below is can be running in a service during a deployment
+  desired_count                      = "${var.desired_count}"
+  deployment_maximum_percent         = "${var.deployment_maximum_percent}"
+  deployment_minimum_healthy_percent = "${var.deployment_minimum_healthy_percent}"
+
+  placement_strategy {
+    type  = "${var.placement_strategy_type}"
+    field = "${var.placement_strategy_field}"
+  }
```
Comment (tamakiii) [inline service_load_balancing/launch_type_ec2.tf]: [NOTE] `aws_ecs_service.placement_strategy` is deprecated
https://www.terraform.io/docs/providers/aws/r/ecs_service.html#placement_strategy

File: service_load_balancing/launch_type_fargate.tf [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  load_balancer {
+    target_group_arn = "${var.target_group_arn}"
+    container_name   = "${var.container_name}"
+    container_port   = "${var.container_port}"
+  }
+
+  lifecycle {
+    # INFO: In the future, we support that U can customize
+    #       https://github.com/hashicorp/terraform/issues/3116
+    ignore_changes = ["desired_count", "task_definition"]
```
Comment (tamakiii) [inline service_load_balancing/launch_type_fargate.tf]: [IMO] Isn't it better adding `task_definition` after the PR is merged?

File: service_load_balancing/README.rst [hunk 4]
```diff
@@ -39,20 +42,40 @@ Basic Usage
...
    }
 
-   # Pre-require your resources declaration (as may be necessary)
    # The following related sources is out of scope on this module
    #resource "aws_alb"              "xxx" {}
    #resource "aws_alb_lister"       "xxx" {}
-   #resource "aws_alb_target_group" "xxx" {}
+   #resource "aws_alb_target_group" "xxx" {
+   #  target_type = "ip" // "ip" for farget, "instance" for the default
+   #}
```
[the remark this answers is not in the document]
Comment (nishigori) [reply service_load_balancing/README.rst]: yes, and require specified "ip" case of fargate 😄

File: service_load_balancing/launch_type_ec2.tf (hunk 2 above)
Comment (nishigori) [reply service_load_balancing/launch_type_ec2.tf]: thx your feedback, I'll fixup but another pulling requests as soon :)

File: service_load_balancing/README.rst [hunk 5]
```diff
@@ -159,6 +176,8 @@ See more details about `Service Auto Scaling`_ in the official AWS docs.
 On ecs-cluster created with our terraform module
 ------------------------------------------------
 
+Maybe only use launch_type as "EC2"
```
[the remark this answers is not in the document]
Comment (nishigori) [reply service_load_balancing/README.rst]: `as` is frequently used on this module  :sweat_smile:

File: service_load_balancing/launch_type_fargate.tf (hunk 3 above)
Comment (nishigori) [reply service_load_balancing/launch_type_fargate.tf]: On this module, currently no-support task_definition updates changing.

We are waiting for https://github.com/hashicorp/terraform/issues/3116
and not yet fix-up ;-(

You can change to container_definition & task parameters via
[ecs_deploy](https://github.com/silinternational/ecs-deploy), AWS CodePipeline, and ecs-cli, ...

but we'll fixup in the future 👍

**Pull request merged**